docker/archive/tar.go

```diff
@@ -0,0 +1,210 @@
+package archive
```

Comment:
(Overall, this is pretty difficult to review, with all the renamings and subtle changes in semantics. In the future, would you consider doing such changes as a series of steps, 1) move original file unchanged + update references, 2) a series of renamings, one commit per one logical change, 3) each change in semantics in one commit, 4) refactorings to move code around, one ide per commit, 5) iterate 234 as necessary)?  Reviewing a dozen small commits with a clear idea and limited scope is much easier than trying to figure out what has changed between 500 lines of the original and completely refactored 500 lines of the new code.)